### PR TITLE
feat(csv): #CRRE-542 add offers to prescriber export + refacto

### PR DESCRIPTION
**Pull request by Sofianebernoussi**
## Describe your changes

- Add offers to export csv for prescriber
- Refacto export 

## Checklist tests
As a prescriber, go to my requests, select a csv in old or not, and export csv to see if you have offers

## Issue ticket number and link
[CRRE-542](https://jira.support-ent.fr/browse/CRRE-542)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)
https://confluence.support-ent.fr/display/CRRE/API+Order

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/crre/helpers/DateHelper.java [hunk 1]
```diff
@@ -33,4 +46,18 @@ public static String convertStringDateToOtherFormat(String stringDate, String ac
...
         return new SimpleDateFormat(newFormat).format(date);
     }
+
+    public static String formatDate(String inputDateStr) {
+        for (String dateFormat : DATE_FORMATS) {
+            try {
+                LocalDate date = LocalDate.parse(inputDateStr, DateTimeFormatter.ofPattern(dateFormat));
+                return DateTimeFormatter.ofPattern(DAY_FORMAT).format(date);
+            } catch (DateTimeParseException e) {
+                // Try next format
```
Comment (flmartineau) [inline src/main/java/fr/openent/crre/helpers/DateHelper.java:56]: Il manque rien ici ?

**Review by flmartineau — APPROVED** [review]

File: src/main/resources/public/ts/model/Utils.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import {Equipment, Offer, Offers} from "./Equipment";
 import {idiom as lang, moment} from "entcore";
 import {Basket, BasketOrder, Baskets} from "./Basket";
-import {OrderClient} from "./OrderClient";
+import {OrderClient, OrdersClient} from "./OrderClient";
```
Comment (JinDouang) [inline src/main/resources/public/ts/model/Utils.ts:4]: OrderClient

et 
import {OrderUniversal} from "./OrderUniversal";

non utilisé

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/crre/helpers/DateHelper.java (hunk 1 above)
Comment (Sofianebernoussi) [reply src/main/java/fr/openent/crre/helpers/DateHelper.java:56]: nop justement on passe juste a la prochaine date

[the remark this answers is not in the document]
Comment (Sofianebernoussi) [reply src/main/java/fr/openent/crre/helpers/DateHelper.java:56]: nop on passe juste a la prochaine date pour avoir le bon format

File: src/main/resources/public/ts/model/OrderUniversal.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ export class OrderUniversal {
     private _equipment_tva5: number;
     private _equipment_tva20: number;
     private _equipment_priceht: number;
-    private _offers: string;
+    private _offers: any[];
```
[the remark this answers is not in the document]
Comment (Sofianebernoussi) [reply src/main/resources/public/ts/model/OrderUniversal.ts:33]: pas le temps malheuresement pour le moment :(

File: src/main/java/fr/openent/crre/controllers/OrderController.java [hunk 4]
```diff
@@ -349,126 +347,161 @@ public void filter(HttpServerRequest request) {
...
                             });
-                        }
-                    });
                 });
     }
 
-    @Get("/orders/old/exports")
-    @ApiDoc("Export list of customer's old orders as CSV")
-    @SecuredAction(value = "", type = ActionType.RESOURCE)
-    @ResourceFilter(PrescriptorAndStructureRight.class)
```
[the remark this answers is not in the document]
Comment (Sofianebernoussi) [reply src/main/java/fr/openent/crre/controllers/OrderController.java:403]: a l'instant, j'ai modif la pr

File: src/main/java/fr/openent/crre/controllers/OrderController.java [hunk 5]
```diff
@@ -349,126 +347,161 @@ public void filter(HttpServerRequest request) {
...
-                            JsonObject order;
-                            for (int i = 0; i < orderClients.size(); i++) {
-                                order = orderClients.getJsonObject(i);
-                                orderMap = new JsonObject();
-                                orderMap.put(Field.NAME, order.getString("equipment_name"));
-                                orderMap.put("ean", order.getString("equipment_key"));
-                                setOrderMap(orders, orderMap, order, null, true);
+    private List<OrderUniversalModel> computeOffersUniversal(OrderUniversalModel orderUniversal) {
+        List<OrderUniversalModel> offers = new ArrayList<>();
+        if (orderUniversal.getOffers().getJsonObject(0).getString(Field.ID) != null) {
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/controllers/OrderController.java:416]: Attention au NPE avec .getJsonObject(0)
tu peux rejouter un if (orderUniversal.getOffers().isEmpty() || orderUniversal.getOffers().getValue(0) instance JsonObject)

File: src/main/java/fr/openent/crre/controllers/OrderController.java [hunk 6]
```diff
@@ -349,126 +347,161 @@ public void filter(HttpServerRequest request) {
...
+                        .setAmount(offer.getInteger(Field.AMOUNT))
+                        .setEquipmentName(offer.getString(Field.NAME))
+                        .setComment(offer.getString(Field.COMMENT))
+                        .setEquipmentKey(offer.getString(Field.EAN))
+                        .setPrescriberValidationDate(offer.getString(Field.CREATION_DATE))
+                        .setBasket(orderUniversal.getBasket());
+                offers.add(orderUniversalOffer);
+            });
+        } else {
+            JsonArray leps = orderUniversal.getOffers().getJsonObject(0).getJsonArray(Field.LEPS);
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/controllers/OrderController.java:428]: Attention au NPE avec .getJsonObject(0)
tu peux rejouter un if (orderUniversal.getOffers().isEmpty() || orderUniversal.getOffers().getValue(0) instance JsonObject)

File: src/main/java/fr/openent/crre/controllers/OrderController.java [hunk 7]
```diff
@@ -349,126 +347,161 @@ public void filter(HttpServerRequest request) {
...
+            }
+        }
+        return offers;
     }
 
-    private void getOrderEquipmentOld
-            (List<Integer> idsOrders, UserInfos user, String idStructure, String idCampaign, String startDate, String endDate, Handler<Either<String, JsonArray>> handlerJsonArray) {
-        orderService.listExport(idsOrders, user, idStructure, idCampaign, null, startDate, endDate, true, handlerJsonArray);
-    }
+    private JsonArray computeOffers(OrderUniversalModel orderUniversalModel, JsonArray equipmentOffers) {
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/controllers/OrderController.java:465]: JsonArray equipmentOffers -> List\<JsonObject\>

File: src/main/java/fr/openent/crre/model/OrderUniversalModel.java [hunk 8]
```diff
@@ -69,7 +72,7 @@ public OrderUniversalModel(JsonObject jsonObject) {
         this.equipmentTva5 = jsonObject.getDouble(Field.EQUIPMENT_TVA5);
         this.equipmentTva20 = jsonObject.getDouble(Field.EQUIPMENT_TVA20);
         this.equipmentPriceht = jsonObject.getDouble(Field.EQUIPMENT_PRICEHT);
-        this.offers = jsonObject.getString(Field.OFFERS);
+        this.offers = jsonObject.getString(Field.OFFERS) != null ? new JsonArray(jsonObject.getString(Field.OFFERS)) : new JsonArray();
```
Comment (BastienBenard) [inline src/main/java/fr/openent/crre/model/OrderUniversalModel.java:75]: Il faudrait mettre un try catch DecodeException e

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 9]
```diff
@@ -160,44 +160,30 @@ export class OrdersClient extends Selection<OrderClient> {
...
+                    let newOffers: Array<Offer> = Mix.castArrayAs(Offer, order.offers);
+                    let offers: Offers = new Offers();
+                    newOffers.forEach((newOffer: Offer) => {
+                        let offer: Offer = new Offer();
+                        offer.name = newOffer.titre;
+                        offer.value = parseInt(newOffer.amount);
+                        offer.id = newOffer.id;
+                        offers.all.push(offer);
+                    });
+                    orderMap.offers = offers;
```
Comment (BastienBenard) [inline src/main/resources/public/ts/model/OrderClient.ts:173]:
```suggestion
                    orderMap.offers = Mix.castArrayAs(Offer, order.offers).map((newOffer: Offer) => {
                        let offer: Offer = new Offer();
                        offer.name = newOffer.titre;
                        offer.value = parseInt(newOffer.amount);
                        offer.id = newOffer.id;
                        return offer
                    });
```

**Pull request merged**